java-checks/src/main/java/org/sonar/java/checks/security/UnpredictableSaltCheck.java

```diff
@@ -0,0 +1,164 @@
+    .ofSubTypes("javax.crypto.spec.PBEKeySpec")
+    .constructor()
+    .withAnyParameters()
+    .build();
+
+  private static final MethodMatchers GET_BYTES = MethodMatchers.create()
+    .ofTypes("java.lang.String")
+    .names("getBytes")
+    .withAnyParameters()
+    .build();
+
+  private final Map<Symbol, Integer> updateCalls = new HashMap<>();
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Arrays.asList(Tree.Kind.METHOD_INVOCATION, Tree.Kind.NEW_CLASS);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (tree.is(Tree.Kind.NEW_CLASS)) {
+      NewClassTree newClassTree = (NewClassTree) tree;
+      checkPBEKeySpec(newClassTree);
+      return;
+    }
+    checkDigestMethod((MethodInvocationTree) tree);
+  }
+
+  @Override
+  public void setContext(JavaFileScannerContext context) {
+    updateCalls.clear();
+    super.setContext(context);
+  }
+
+  private void checkDigestMethod(MethodInvocationTree methodInvocationTree) {
+    if (UPDATE_MESSAGE_DIGEST.matches(methodInvocationTree)) {
+      getInvokedSymbol(methodInvocationTree).ifPresent(symbol -> 
+        updateCalls.compute(symbol, (k, v) -> (v == null) ? 1 : (v + 1)));
+    } else if (DIGEST_METHOD.matches(methodInvocationTree)) {
+      getInvokedSymbol(methodInvocationTree).ifPresent(symbol -> {
+        Integer count = updateCalls.get(symbol);
+        if (count == null || (count == 1 && methodInvocationTree.arguments().isEmpty())) {
+          reportIssue(methodInvocationTree, ADD_SALT);
+        }
+      });
+    }
+  }
+
+  private void checkPBEKeySpec(NewClassTree newClassTree) {
+    if (NEW_PBE_KEY_SPEC.matches(newClassTree)) {
+      if (newClassTree.arguments().size() <= 1) {
+        reportIssue(newClassTree, ADD_SALT);
+      } else {
+        ExpressionTree saltExpression = ExpressionUtils.skipParentheses(newClassTree.arguments().get(1));
+        if (isPredictable(saltExpression)) {
+          reportIssue(newClassTree, UNPREDICTABLE_SALT);
+        }
+      }
+    }
+  }
+
+  private static boolean isPredictable(ExpressionTree saltExpression) {
+    return (saltExpression.is(Tree.Kind.METHOD_INVOCATION) && isInitializedWithGetBytes((MethodInvocationTree) saltExpression)) ||
+      (saltExpression.is(Tree.Kind.IDENTIFIER) && isInitializedWithLiteral((IdentifierTree) saltExpression));
+  }
+  
+  private static Optional<Symbol> getInvokedSymbol(MethodInvocationTree mit) {
+    ExpressionTree methodSelect = mit.methodSelect();
+    if (methodSelect.is(Tree.Kind.MEMBER_SELECT)) {
+      ExpressionTree expression = ((MemberSelectExpressionTree) methodSelect).expression();
+      if (expression.is(Tree.Kind.IDENTIFIER)) {
+        return Optional.of(((IdentifierTree) expression).symbol());
+      }
+    }
+    return Optional.empty();
+  }
+
+  private static boolean isInitializedWithLiteral(IdentifierTree identifier) {
+    Tree declaration = identifier.symbol().declaration();
```

Comment:
It's a corner case, but you should check that the variable is not re-assigned, otherwise you could be subject to FP.
You can probably use `symbol.isFinal() || JUtils.isEffectivelyFinal(symbol)` to do this.